Avoid clobbering environment variables loaded from /etc/profile.d

ArrayBolt3 opened the pull request:
qubes-session loads environment variables from
`systemctl --user show-environment`, so that variables set using systemd environment generators are present in the user session. Previously, when one of those generators provided an environment variable that was already set by a script under /etc/profile.d, the variable from the environment generator would clobber the variable from /etc/profile.d. This is backwards - variables from /etc/profile.d should take precedence over those defined in systemd environment generators.

To fix this, update the systemd user manager's environment with the contents of the session's environment immediately before updating the session's environment with the environment from the systemd user manager. This two-way sync means that all variables defined in both environment generators and profile scripts end up in the user's environment, while making variable definitions from the profile scripts take priority over variable definitions from environment generators.

Fixes: https://github.com/QubesOS/qubes-issues/issues/10299

ArrayBolt3 commented:
Note that this results in a lot of new environment variables being loaded into systemd's user manager that weren't previously. I believe this is correct behavior, however it might change the behavior of other user services and qrexec services. Whatever testing needs to be done to make sure that doesn't break the world, should probably be done.

marmarek commented:
This will now break some of the systemd-defined variables (including XDG_DATA_DIRS parts added there, so - flatpak). The problem with this approach is that /etc/profile.d scripts are usually handling already existing variables by appending value (that applies at least to PATH, but XDG_DATA_DIRS too). If you now override one with the other, you break /etc/profile.d expectation (it was meant to extend values, but now it will override them). What probably wants happening here, is loading systemd-defined variables earlier, before /etc/profile gets loaded. I'm not 100% sure where that actually should be. Maybe somewhere in https://github.com/QubesOS/qubes-gui-agent-linux/blob/main/gui-agent/qubes-gui-runuser.c (which is used in https://github.com/QubesOS/qubes-gui-agent-linux/blob/main/appvm-scripts/usrbin/qubes-run-xorg) ?

Note also that /etc/profile.d is meant for shell variables. I don't think any spec mandates loading it for non-shell apps. So, if you want a variable to be set for various GUI apps (not just shell running in a terminal emulator), you should probably use a different mechanism (likely /etc/environment.d). For example on native Xfce, if you launch an app from the menu, it won't see variables set via /etc/profile.d (you can check via `/proc/PID/environ`, as ofc opening shell will load /etc/profile.d; check for example LESSOPEN - this one seems to be set only in /etc/profile.d). /etc/profile.d works great for power users who open all applications from a terminal, which is likely why it's such a common habit...

See also https://github.com/QubesOS/qubes-issues/issues/6408 which demanded /etc/environment.d to be respected. And also https://github.com/QubesOS/qubes-gui-agent-linux/pull/151 for some more background.

ArrayBolt3 commented:
Maybe moving to environment.d in the future can be a goal, but right now to my awareness *all* of the system-wide environment variable manipulation in Whonix (and quite a bit of it done in Qubes it appears) is done in /etc/profile.d or /etc/X11/Xsession.d, and Whonix does a lot of environment variable manipulation. Changing everything to environment.d this late in the game might not be entirely out of the question, but it would definitely fall into the "last resort" category.

Loading things from /etc/environment.d first sounds like a good solution. qubes-gui-runuser sounds like kind of an awkward place to do it though since that requires lots of possibly messy string parsing in C. It seems like it would be easier (possibly) to do it in all scripts that eventually end up calling qubes-session. I don't know where all it is called though, and a string search in Github [wasn't all that clear](https://github.com/search?q=org%3AQubesOS+qubes-session&type=code&p=1).

ArrayBolt3 commented:
I guess after further thought, maybe implementing this in qubes-gui-runser.c is better. Otherwise we risk this bug popping back up if someone creates a new script that calls qubes-session via qubes-gui-runuser and doesn't parse the environment variables from systemd. The environment variables are just going to be newline-separated key-value pairs, so whatever "parsing" is needed can't be that hard.

marmarek commented:
I proposed qubes-gui-runuser, because that's what calls shell that will eventually call qubes-session, so it runs before shell startup scripts.
As for parsing, I don't think you really need it. You can simply ask systemd for the variables using proper API (dbus), no need for asking `systemctl --user show-environment` to format it into string just to have it split into variables back. It will add dbus dependency to that tool, but it's kinda there already via pam_systemd. Oh, maybe there is a simpler solution if there is some pam module that would load those variables?

ArrayBolt3 commented:
There's code in qubes-gui-runuser that can work without PAM, it seems to me like a bugfix should ideally work in both scenarios. Probably better to use dbus.

ArrayBolt3 committed:
qubes-gui-runner: get environment from systemd user instance and pass it through to the session

qubes-session previously loaded environment variables from
`systemctl --user show-environment`, so that variables set using systemd
environment generators would be present in the user session. However,
doing this meant any environment variables defined or augmented by
scripts under /etc/profile.d could be clobbered by versions of those
same variables from the systemd user instance, resulting in application
misbehavior in some instances.

To fix this, load all environment variables from systemd's user instance
in qubes-gui-runner instead. This will augment the environment provided
by PAM with the environment provided by systemd, then the /etc/profile.d
scripts can augment the environment further. This should prevent any
variables from being incorrectly clobbered, and allow all mechanisms of
providing environment variables to the end-user's session to function
as intended.

Fixes: https://github.com/QubesOS/qubes-issues/issues/10299

ArrayBolt3 commented:
Well, turns out when https://dbus.freedesktop.org/doc/api/html/index.html says "If you use this low-level API directly, you're signing up for some pain", it is not joking! My latest commit is entirely untested, probably won't even compile, and doesn't handle the non-PAM case I wanted it to handle, but, for whatever it's worth, it's there. If this is an entirely unsuitable approach, I can give up on this, otherwise I'll finish polishing and testing this in the near future.

ArrayBolt3 commented on gui-agent/qubes-gui-runuser.c:
```diff
@@ -32,13 +32,138 @@
+    env_arr[env_arr_len - 1] = NULL;
+
+dbus_cleanup:
+    if (dbus_conn != NULL)
+        dbus_connection_unref(dbus_conn);
+    if (env_request != NULL)
+        dbus_message_unref(env_request);
+    if (env_reply != NULL)
+        dbus_message_unref(env_reply);
+    if (inner_iter_typesig != NULL)
+        dbus_free(inner_iter_typesig);
```
Good point.

ArrayBolt3 commented:
There are helpers for this, but they are mainly parts of larger application frameworks like (quoting from the API's main page) GLib, Qt, Python, Mono, Java, etc. D-Bus is also intended to be an event-driven RPC framework where things are managed by callback functions and a main loop, whereas what I'm doing is more like calling an HTTP endpoint and getting a response back. That isn't quite the normal use case for D-Bus AFAICT, thus I don't expect there will be a well-supported helper in general for this kind of thing.

Good point about pam_open_session, I'll move the call appropriately (and hopefully fully test this before the next push).

ArrayBolt3 commented:
It looks like the no-PAM code in qubes-gui-runuser is entirely broken - the `execve` line at the end references a variable `env` that doesn't even exist in the function or as a global. Not sure if it's worth trying to fix it now, I'll leave it broken and not add the systemd environment fixes. In the long run, should the code be repaired, or would it be better to just discard it?

ArrayBolt3 commented:
Alright, this time it builds locally, and it actually seems to work on my end, though I've not tested it nearly as much needed to consider it rigorously tested. I think at this point we can be fairly certain this solution *can* work, so some security review on the mess required to talk to systemd over D-Bus would be much appreciated.

Some TODOs that aren't already in the code:

* The error messages from the D-Bus interaction code should make their origin more clear, and it's possible the errors themselves could be clearer.
* There should be a comment somewhere documenting that one should never pass a `char **` that contains any unsafe-to-free memory to `get_systemd_env_from_dbus`. The function reallocarray's the passed-in array numerous times, and can `free` elements within the array to replace them with elements copied from systemd.
* This whole mechanism should be tested in isolation - right now all I know is my environment looks like I'd expect (without things from /etc/profile.d being clobbered) when I boot with this, I don't yet know if getting the environment variables from systemd actually succeeded or not.

ArrayBolt3 commented on gui-agent/qubes-gui-runuser.c:
```diff
@@ -31,14 +31,243 @@
+            err(1, "Failed to allocate memory for environment array");
+        env_arr[env_arr_len - 1] = strdup(env_val);
+        if (env_arr[env_arr_len - 1] == NULL)
+            err(1, "Failed to allocate memory for environment item");
```
Might help to actually walk through the list rather than just spinning on the first element... The fact that this didn't crash my VM makes me believe this is probably failing to connect to systemd at all (might be a timing issue).

```suggestion
            err(1, "Failed to allocate memory for environment item");
            
            dbus_message_iter_next(&reply_arr_iter);
```

marmarek commented:
Yes, it looks related. It's likely due to this part of qubes-run-xorg:
```
if qsvc guivm-gui-agent; then
    DISPLAY_XORG=:1

    # Create Xorg. Xephyr will be started using qubes-start-xephyr later.
    exec runuser -u "$DEFAULT_USER" -- /bin/sh -l -c "exec /usr/lib/qubes/qubes-xorg-wrapper $DISPLAY_XORG -nolisten tcp vt07 -wr -config xorg-qubes.conf > ~/.xorg-errors 2>&1"
else
    # Use sh -l here to load all session startup scripts (/etc/profile, ~/.profile
    # etc) to populate environment. This is the environment that will be used for
    # all user applications and qrexec calls.
    exec /usr/bin/qubes-gui-runuser "$DEFAULT_USER" /bin/sh -l -c "exec /usr/bin/xinit $XSESSION -- /usr/lib/qubes/qubes-xorg-wrapper :0 -nolisten tcp vt07 -wr -config xorg-qubes.conf > ~/.xsession-errors 2>&1"
fi
```
sys-gui does use the guivm-gui-agent branch.
So, it uses `runuser` instead of `qubes-gui-runuser`. While PAM configuration includes pam_systemd, and logs say user's systemd instance was started, I don't see session bus started. And this is actually expected, because this X instance (`:1`) is only a thin wrapper to forward a single full-screen window to dom0, actual session bus is started later, under the other X server instance (Xephyr started later, as `:0`).
So, the fix might be as easy as checking if `DBUS_SESSION_BUS_ADDRESS` variable is set first.

FWIW the test VM is running, with sys-gui created. It's enough to start it and then check `qubes-gui-agent` service status (`qvm-run -p --nogui sys-gui systemctl status qubes-gui-agent`). You can also logout from dom0 and select GUI domain as session type in lightdm. But be careful with screenlocker in sys-gui, as user password is not set there, if you let it lock, you'd need to switch to tty2 and kill it via qvm-run or so.

ArrayBolt3 commented:
Well this is curious, because I saw that same error when I tested this on debian-13-xfce yesterday. I ignored it because everything seemed to work anyway, I figured it must have fallen back to some other way of getting the info or maybe the error was unrelated. I'll try this solution and see if checking for a non-empty `DBUS_SESSION_BUS_ADDRESS` does the trick. Either way, I think the `|| exit` in 00-qubes-systemd-env-import.sh should change to an `|| true`. If this fails, there's no reason to block further processing of `profile.d` or similar.

I'm in the VM and testing now.
